filebeat/input/filestream/filestream.go

```diff
@@ -83,16 +86,10 @@ func newFileReader(
 		offset:             offset,
 		lastTimeRead:       time.Now(),
 		backoff:            backoff.NewExpBackoff(canceler.Done(), config.Backoff.Init, config.Backoff.Max),
-		tg:                 unison.TaskGroup{},
+		readerCtx:          readerCtx,
+		tg:                 tg,
 	}
 
-	l.ctx, l.cancelReading = ctxtool.WithFunc(ctxtool.FromCanceller(canceler), func() {
-		err := l.tg.Stop()
-		if err != nil {
-			l.log.Errorf("Error while stopping filestream logFile reader: %v", err)
-		}
-	})
```

Comment:
The group now shuts down when the `readerCtx` gets cancelled. This is a slight change in semantics, as the reader context was closed in the past only after the taskgroup go-routines have finished (Stop used to wait).

 In order to make behavior and signal propagation a little more predictable, the change ensures that we have a clear chain of cancellation contexts:
`canceler -> readerCtx -> (taskgroup ctx)`. Before we used to have something like:

```
canceler -> ctx
canceler -> (taskgroup ctx)
```